chips/cc26x2/src/rfc.rs

```diff
@@ -0,0 +1,580 @@
+            status = self.cmdsta();
+            if (status & 0xFF) == 0x01 {
+                return Ok(());
+            }
+
+            timeout += 1;
+        }
+
+        return Err(status);
+    }
+
+    fn wait_cmdr(&self, rf_command: u32) -> RfcResult {
+        let command_op: &cmd::CmdCommon = unsafe { &*(rf_command as *const cmd::CmdCommon) };
+
+        let mut timeout: u32 = 0;
+        let mut status = 0;
+        const MAX_TIMEOUT: u32 = 0x2FFFFFF;
+        while timeout < MAX_TIMEOUT {
+            status = command_op.status.get();
+            if status == 0x0400 {
+                return Ok(());
+            }
+
+            timeout += 1;
+        }
+
+        return Err(status as u32);
+    }
+
+    fn cmdsta(&self) -> u32 {
+        let dbell_regs = RFC_DBELL_BASE;
+        let ret: u32 = dbell_regs.cmdsta.get();
+
+        return ret;
+    }
+
+    fn enable_hw_interrupts(&self) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Enable all interrupts
+        dbell_regs
+            ._rfhwien
+            .modify(RFHWInterrupts::ALL_INTERRUPTS::SET);
+    }
+
+    pub fn handle_hw_interrupts(&self) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Clear all RFHW interrupts
+        dbell_regs
+            ._rfhwifg
+            .write(RFHWInterrupts::ALL_INTERRUPTS::SET);
+    }
+
+    fn disable_hw_interrupts(&self) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Disable all RFHW interrupts
+        dbell_regs
+            ._rfhwien
+            .modify(RFHWInterrupts::ALL_INTERRUPTS::CLEAR);
+        // Clear all RFHW interrupts
+        dbell_regs
+            ._rfhwifg
+            .write(RFHWInterrupts::ALL_INTERRUPTS::SET);
+    }
+
+    fn enable_cpe_interrupts(&self) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Enable CPE interrupts
+        dbell_regs
+            .rfcpeien
+            .modify(CPEInterrupts::ALL_INTERRUPTS::SET);
+    }
+
+    pub fn handle_cpe_interrupts(&self) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Clear all CPE interrupts
+        dbell_regs.rfcpeifg.set(0x7FFFFFFF);
+    }
+
+    fn disable_cpe_interrupts(&self) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Disable all CPE interrupts
+        dbell_regs
+            .rfcpeien
+            .modify(CPEInterrupts::ALL_INTERRUPTS::CLEAR);
+        // Clear all CPE interrupts
+        dbell_regs.rfcpeifg.set(0x7FFFFFFF);
+    }
+
+    pub fn cpe_vec_select(&self, cpe: bool) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Select CPE0 or CPE1
+        if !cpe {
+            dbell_regs.rfcpeisl.modify(CPEVectorSelect::ALL::CLEAR);
+        } else {
+            dbell_regs.rfcpeisl.modify(CPEVectorSelect::ALL::SET);
+        }
+    }
+
+    pub fn handle_ack_interrupt(&self) {
+        let dbell_regs = RFC_DBELL_BASE;
+        // Reset flag for Command ACK
+        dbell_regs.rfackifg.write(DBellCmdAck::CMDACK::SET);
+    }
+
+    pub fn send<T>(&self, rf_command: &T) -> RfcResult {
```

Comment:
There are three big problems with this signature.

  1. It lets a client effectively serialize and send _any_ data structure to the radio, including it's private fields. That's because `T` isn't constrained in any way. So imagine a client has a reference to an encryption capsule which, internally, contains a secret encryption key. This method allows that client to serialize the entire capsule, including the encryption key to the radio. Preferably the argument would be either an already serialized type (e.g. a slice), or a particular set of types (e.g. that all implement a `RadioCommand` trait or something).

  2. The lifetime of the argument is unconstrained, meaning there is no guarantee the memory for the underlying value isn't de-allocated as soon as `send` returns. Instead, the argument needs to have `'static` lifetime, since we don't know (at the language level anyway) when the radio is going to be done using it.

  3. In some cases, the memory is going to be mutated by the radio (e.g. `CMD_COUNT_BRANCH` or `CMD_READ_RFREG`) so in order to avoid hitting undefined behavior, this needs to be a unique `&mut` reference (or potentially a type that has interior mutability in all the fields containing memory the radio might modify).

So the high bit is that right now, `send` would let a client circumvent the Rust type system in three ways. The root of the problem is in the definition of the register, but that's gonna be harder to fix (we'll fix it, just potentially not in this PR).

For now, the simplest solution might be to accept a serialized command as a `&[u8]`, and copy it into a statically allocated buffer that's as large as the largest command.

Reply:
from out-of-thread discussion: These could be private until the underlying issue with cmdr is fixed. Really the user should not be writing anything outside of the commands to cmdr which they should only be able to invoke through commands.rs and the public fn specifically for setup, powerdown, and rx/tx ops for now. Not the ideal fix.